Cumulative Incidence (Kaplan-Meier) for competing risks.

mvlvrd opened the pull request:
Add non-parametric Cumulative Incidence (Kaplan-Meier) for competing risks. Tests included.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://scikit-survival.readthedocs.io/en/latest/contributing.html#making-changes-to-the-code
-->

**Checklist**
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] closes #490
- [x] pytest passes
- [x] tests are included
- [x] code is well formatted
- [x] documentation renders correctly

**What does this implement/fix? Explain your changes**
Implements cumulative incidence estimator for the competing risks case. Confidence Intervals are not implemented but are under development.

mvlvrd commented:
The codacy linter triggers for unused unpacked assignments, however the ruff linter doesn't. Perhaps it would be nice to make them consistent, though is not a big issue. (I can open a new issue/PR if needed, it is trivial to fix it from the ruff side).

sebp reviewed (changes requested):
This is great, thanks a lot!

The code seems to be correct to me. I would suggest to improve the API docs a little bit to avoid ambiguities.

In addition, I'm not sure if certain corner cases are covered. For instance, what happens if no event has been observed for one (or more) competing risk?

I would appreciate if a test case with more than 2 competing risks could be added.

mvlvrd commented:
Thanks so much for the review!
I think I have addressed most of your points. I have been looking for a real dataset with more than two risks, but I couldn't find one, so I added a simulated one. Corner cases like zero events for one (or more) particular risk are not included and raise an error.

I also changed the function name (removing the reference to Kaplan-Meier). This is because I found a couple of papers that use KM for the naive extension of KM to the competing risks case, while we are using the rigorous estimator that does not assume independence between risks.

sebp commented on sksurv/nonparametric.py:
```diff
@@ -586,3 +599,117 @@ def predict_ipcw(self, y):
+    >>>    plt.step(x, y[i], where="post", label=i)
+    >>> plt.ylim(0, 1)
+    >>> plt.legend()
+    >>> plt.show()
```
CSV if fine. Can you please add an entry to https://github.com/sebp/scikit-survival/blob/master/sksurv/datasets/data/README.md describing the dataset with a link to its source. Adding a function to load the dataset to `sksurv.datasets` would be great.

sebp commented on tests/data/bmt.csv:
Can you please remove this file and use sksurv.datasets.load_bmt instead
